[ICE]: `found DefPathHash collision between DefPath`

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.
http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/
-->

### Code
This first ICEs and then hangs while printing the query stack

```Rust
impl Iterator {
  reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:
```
rustc 1.96.0-nightly (d9563937f 2026-03-03)
binary: rustc
commit-hash: d9563937fa3b030c5845811113505070109414d2
commit-date: 2026-03-03
host: x86_64-unknown-linux-gnu
release: 1.96.0-nightly
LLVM version: 22.1.0
```

### Error output

```
<output>
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error[E0575]: expected method or associated constant, found associated type `Iterator::Item`
 --> a.rs:2:8
  |
2 |   reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
  |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a method or associated constant

error[E0658]: functions delegation is not yet fully implemented
 --> a.rs:2:3
  |
2 |   reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
  |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: see issue #118212 <https://github.com/rust-lang/rust/issues/118212> for more information
  = help: add `#![feature(fn_delegation)]` to the crate attributes to enable
  = note: this compiler was built on 2026-03-03; consider upgrading it if it is out of date


thread 'rustc' (4021229) panicked at /rustc-dev/d9563937fa3b030c5845811113505070109414d2/compiler/rustc_hir/src/definitions.rs:70:13:
found DefPathHash collision between DefPath {
    data: [
        DisambiguatedDefPathData {
            data: Impl,
            disambiguator: 0,
        },
        DisambiguatedDefPathData {
            data: ValueNs(
                "partition_in_place",
            ),
            disambiguator: 0,
        },
        DisambiguatedDefPathData {
            data: LifetimeNs(
                "'a",
            ),
            disambiguator: 0,
        },
    ],
    krate: crate0,
} and DefPath {
    data: [
        DisambiguatedDefPathData {
            data: Impl,
            disambiguator: 0,
        },
        DisambiguatedDefPathData {
            data: ValueNs(
                "partition_in_place",
            ),
            disambiguator: 0,
        },
        DisambiguatedDefPathData {
            data: LifetimeNs(
                "'a",
            ),
            disambiguator: 0,
        },
    ],
    krate: crate0,
}. Compilation cannot continue.
stack backtrace:
   0:     0x7f14dad6495b - <<std[f1c45ce4ee161762]::sys::backtrace::BacktraceLock>::print::DisplayBacktrace as core[bca3ab08014da8dd]::fmt::Display>::fmt
   1:     0x7f14db4228c8 - core[bca3ab08014da8dd]::fmt::write
   2:     0x7f14dad7b966 - <std[f1c45ce4ee161762]::sys::stdio::unix::Stderr as std[f1c45ce4ee161762]::io::Write>::write_fmt
   3:     0x7f14dad3aa38 - std[f1c45ce4ee161762]::panicking::default_hook::{closure#0}
   4:     0x7f14dad57dc3 - std[f1c45ce4ee161762]::panicking::default_hook
   5:     0x7f14d9d43eec - std[f1c45ce4ee161762]::panicking::update_hook::<alloc[69236e05dd7e86e4]::boxed::Box<rustc_driver_impl[87c87f3e97b6922c]::install_ice_hook::{closure#1}>>::{closure#0}
   6:     0x7f14dad580a2 - std[f1c45ce4ee161762]::panicking::panic_with_hook
   7:     0x7f14dad3aaf8 - std[f1c45ce4ee161762]::panicking::panic_handler::{closure#0}
   8:     0x7f14dad31829 - std[f1c45ce4ee161762]::sys::backtrace::__rust_end_short_backtrace::<std[f1c45ce4ee161762]::panicking::panic_handler::{closure#0}, !>
   9:     0x7f14dad3c55d - __rustc[e7c1d660df8da2b8]::rust_begin_unwind
  10:     0x7f14d7917ddc - core[bca3ab08014da8dd]::panicking::panic_fmt
  11:     0x7f14dbccb465 - <rustc_middle[5922741a8fb3ae5b]::ty::context::TyCtxt>::create_def
  12:     0x7f14d986b64c - <<rustc_ast_lowering[2f0d4b4a36627f7a]::delegation::generics::HirOrAstGenerics>::into_hir_generics::{closure#0} as core[bca3ab08014da8dd]::ops::function::FnOnce<(&mut rustc_ast[5f0b67c0ef4a6e26]::ast::Generics,)>>::call_once
  13:     0x7f14d98527ed - <rustc_ast_lowering[2f0d4b4a36627f7a]::delegation::generics::HirOrAstGenerics>::into_hir_generics
  14:     0x7f14d9849f44 - <rustc_ast_lowering[2f0d4b4a36627f7a]::LoweringContext>::process_segment
  15:     0x7f14d984caf0 - <rustc_ast_lowering[2f0d4b4a36627f7a]::LoweringContext>::lower_delegation
  16:     0x7f14dbba50c1 - <rustc_ast_lowering[2f0d4b4a36627f7a]::item::ItemLowerer>::lower_node
  17:     0x7f14dbb9bd67 - rustc_ast_lowering[2f0d4b4a36627f7a]::lower_to_hir
  18:     0x7f14dc5d5368 - rustc_query_impl[7668830077db65c5]::query_impl::hir_crate::invoke_provider_fn::__rust_begin_short_backtrace
  19:     0x7f14dc5d4244 - rustc_query_impl[7668830077db65c5]::execution::try_execute_query::<rustc_middle[5922741a8fb3ae5b]::query::caches::SingleCache<rustc_middle[5922741a8fb3ae5b]::query::erase::ErasedData<[u8; 8usize]>>, false>
  20:     0x7f14dc5d3ccf - rustc_query_impl[7668830077db65c5]::query_impl::hir_crate::execute_query_non_incr::__rust_end_short_backtrace
  21:     0x7f14db56defd - rustc_query_impl[7668830077db65c5]::query_impl::opt_hir_owner_nodes::invoke_provider_fn::__rust_begin_short_backtrace
  22:     0x7f14db56d0ca - rustc_query_impl[7668830077db65c5]::execution::try_execute_query::<rustc_data_structures[d0703324dfa1e5f8]::vec_cache::VecCache<rustc_span[cd7b03fc0c8acb50]::def_id::LocalDefId, rustc_middle[5922741a8fb3ae5b]::query::erase::ErasedData<[u8; 8usize]>, rustc_middle[5922741a8fb3ae5b]::dep_graph::graph::DepNodeIndex>, false>
  23:     0x7f14db56cd45 - rustc_query_impl[7668830077db65c5]::query_impl::opt_hir_owner_nodes::execute_query_non_incr::__rust_end_short_backtrace
  24:     0x7f14dbcfaee6 - <rustc_middle[5922741a8fb3ae5b]::ty::context::TyCtxt>::hir_owner_node
  25:     0x7f14dc5da911 - rustc_middle[5922741a8fb3ae5b]::hir::map::hir_crate_items
  26:     0x7f14dc5d5f31 - rustc_query_impl[7668830077db65c5]::query_impl::hir_crate_items::invoke_provider_fn::__rust_begin_short_backtrace
  27:     0x7f14dc5d4244 - rustc_query_impl[7668830077db65c5]::execution::try_execute_query::<rustc_middle[5922741a8fb3ae5b]::query::caches::SingleCache<rustc_middle[5922741a8fb3ae5b]::query::erase::ErasedData<[u8; 8usize]>>, false>
  28:     0x7f14dc5d3a4f - rustc_query_impl[7668830077db65c5]::query_impl::hir_crate_items::execute_query_non_incr::__rust_end_short_backtrace
  29:     0x7f14db569a18 - rustc_interface[a32636ddb3a490e3]::passes::analysis
  30:     0x7f14dc5d2fdc - rustc_query_impl[7668830077db65c5]::execution::try_execute_query::<rustc_middle[5922741a8fb3ae5b]::query::caches::SingleCache<rustc_middle[5922741a8fb3ae5b]::query::erase::ErasedData<[u8; 0usize]>>, false>
  31:     0x7f14dc5d2d3b - rustc_query_impl[7668830077db65c5]::query_impl::analysis::execute_query_non_incr::__rust_end_short_backtrace
  32:     0x7f14dc602aed - rustc_interface[a32636ddb3a490e3]::interface::run_compiler::<(), rustc_driver_impl[87c87f3e97b6922c]::run_compiler::{closure#0}>::{closure#1}
  33:     0x7f14dc5e71be - std[f1c45ce4ee161762]::sys::backtrace::__rust_begin_short_backtrace::<rustc_interface[a32636ddb3a490e3]::util::run_in_thread_with_globals<rustc_interface[a32636ddb3a490e3]::util::run_in_thread_pool_with_globals<rustc_interface[a32636ddb3a490e3]::interface::run_compiler<(), rustc_driver_impl[87c87f3e97b6922c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>
  34:     0x7f14dc5e7820 - <std[f1c45ce4ee161762]::thread::lifecycle::spawn_unchecked<rustc_interface[a32636ddb3a490e3]::util::run_in_thread_with_globals<rustc_interface[a32636ddb3a490e3]::util::run_in_thread_pool_with_globals<rustc_interface[a32636ddb3a490e3]::interface::run_compiler<(), rustc_driver_impl[87c87f3e97b6922c]::run_compiler::{closure#0}>::{closure#1}, ()>::{closure#0}, ()>::{closure#0}::{closure#0}, ()>::{closure#1} as core[bca3ab08014da8dd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  35:     0x7f14dc5e86ac - <std[f1c45ce4ee161762]::sys::thread::unix::Thread>::new::thread_start
  36:     0x7f14d60a598b - <unknown>
  37:     0x7f14d6129a0c - <unknown>
  38:                0x0 - <unknown>

error: the compiler unexpectedly panicked. This is a bug

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: please make sure that you have updated to the latest nightly

note: please attach the file at `/tmp/im/rustc-ice-2026-03-04T15_38_58-4021211.txt` to your bug report

note: rustc 1.96.0-nightly (d9563937f 2026-03-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [hir_crate] getting the crate HIR
```

</p>
</details>

--- a/compiler/rustc_ast_lowering/src/delegation/generics.rs
+++ b/compiler/rustc_ast_lowering/src/delegation/generics.rs
@@ -295,10 +295,9 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
 
             let param_ident = Ident::new(p.name, span);
             let def_name = Some(param_ident.name);
-            let path_data = def_kind.def_path_data(def_name);
             let node_id = self.next_node_id();
 
-            let def_id = self.create_def(node_id, def_name, def_kind, path_data, span);
+            let def_id = self.create_def(node_id, def_name, def_kind, span);
 
             let kind = match p.kind {
                 GenericParamDefKind::Lifetime => {

--- a/compiler/rustc_ast_lowering/src/expr.rs
+++ b/compiler/rustc_ast_lowering/src/expr.rs
@@ -8,7 +8,6 @@ use rustc_data_structures::stack::ensure_sufficient_stack;
 use rustc_errors::msg;
 use rustc_hir as hir;
 use rustc_hir::def::{DefKind, Res};
-use rustc_hir::definitions::DefPathData;
 use rustc_hir::{HirId, Target, find_attr};
 use rustc_middle::span_bug;
 use rustc_middle::ty::TyCtxt;
@@ -472,13 +471,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         for (idx, arg) in args.iter().cloned().enumerate() {
             if legacy_args_idx.contains(&idx) {
                 let node_id = self.next_node_id();
-                self.create_def(
-                    node_id,
-                    None,
-                    DefKind::AnonConst,
-                    DefPathData::LateAnonConst,
-                    f.span,
-                );
+                self.create_def(node_id, None, DefKind::AnonConst, f.span);
                 let const_value =
                     if let ControlFlow::Break(span) = WillCreateDefIdsVisitor.visit_expr(&arg) {
                         Box::new(Expr {

--- a/compiler/rustc_ast_lowering/src/lib.rs
+++ b/compiler/rustc_ast_lowering/src/lib.rs
@@ -51,7 +51,7 @@ use rustc_data_structures::tagged_ptr::TaggedRef;
 use rustc_errors::{DiagArgFromDisplay, DiagCtxtHandle};
 use rustc_hir::def::{DefKind, LifetimeRes, Namespace, PartialRes, PerNS, Res};
 use rustc_hir::def_id::{CRATE_DEF_ID, LOCAL_CRATE, LocalDefId};
-use rustc_hir::definitions::{DefPathData, DisambiguatorState};
+use rustc_hir::definitions::PerParentDisambiguatorState;
 use rustc_hir::lints::{AttributeLint, DelayedLint};
 use rustc_hir::{
     self as hir, AngleBrackets, ConstArg, GenericArg, HirId, ItemLocalMap, LifetimeSource,
@@ -94,7 +94,7 @@ pub mod stability;
 struct LoweringContext<'a, 'hir, R> {
     tcx: TyCtxt<'hir>,
     resolver: &'a mut R,
-    disambiguator: DisambiguatorState,
+    disambiguator: PerParentDisambiguatorState,
 
     /// Used to allocate HIR nodes.
     arena: &'hir hir::Arena<'hir>,
@@ -159,7 +159,7 @@ impl<'a, 'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'a, 'hir, R> {
         Self {
             tcx,
             resolver,
-            disambiguator: DisambiguatorState::new(),
+            disambiguator: Default::default(),
             arena: tcx.hir_arena,
 
             // HirId handling.
@@ -302,6 +302,10 @@ impl<'a, 'tcx> ResolverAstLoweringExt<'tcx> for ResolverDelayedAstLowering<'a, '
     fn next_node_id(&mut self) -> NodeId {
         next_node_id(&mut self.next_node_id)
     }
+
+    fn steal_or_create_disambiguator(&self, parent: LocalDefId) -> PerParentDisambiguatorState {
+        self.base.steal_or_create_disambiguator(parent)
+    }
 }
 
 fn next_node_id(current_id: &mut NodeId) -> NodeId {
@@ -403,6 +407,10 @@ impl<'tcx> ResolverAstLowering<'tcx> {
     #[inline]
     fn next_node_id(&mut self) -> NodeId {
         next_node_id(&mut self.next_node_id)
+    }
+
+    fn steal_or_create_disambiguator(&self, parent: LocalDefId) -> PerParentDisambiguatorState {
+        self.per_parent_disambiguators.get(&parent).map(|s| s.steal()).unwrap_or_default()
     }
 }
 
@@ -717,7 +725,6 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         node_id: ast::NodeId,
         name: Option<Symbol>,
         def_kind: DefKind,
-        def_path_data: DefPathData,
         span: Span,
     ) -> LocalDefId {
         let parent = self.current_hir_id_owner.def_id;
@@ -733,7 +740,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         let def_id = self
             .tcx
             .at(span)
-            .create_def(parent, name, def_kind, Some(def_path_data), &mut self.disambiguator)
+            .create_def(parent, name, def_kind, None, &mut self.disambiguator)
             .def_id();
 
         debug!("create_def: def_id_to_node_id[{:?}] <-> {:?}", def_id, node_id);
@@ -774,6 +781,8 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
     ) {
         let owner_id = self.owner_id(owner);
 
+        let new_disambig = self.resolver.steal_or_create_disambiguator(owner_id.def_id);
+        let disambiguator = std::mem::replace(&mut self.disambiguator, new_disambig);
         let current_attrs = std::mem::take(&mut self.attrs);
         let current_bodies = std::mem::take(&mut self.bodies);
         let current_define_opaque = std::mem::take(&mut self.define_opaque);
@@ -808,6 +817,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         assert!(self.impl_trait_bounds.is_empty());
         let info = self.make_owner_info(item);
 
+        self.disambiguator = disambiguator;
         self.attrs = current_attrs;
         self.bodies = current_bodies;
         self.define_opaque = current_define_opaque;
@@ -1015,7 +1025,6 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
                     param,
                     Some(kw::UnderscoreLifetime),
                     DefKind::LifetimeParam,
-                    DefPathData::DesugaredAnonymousLifetime,
                     ident.span,
                 );
                 debug!(?_def_id);
@@ -2504,13 +2513,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
             // We're lowering a const argument that was originally thought to be a type argument,
             // so the def collector didn't create the def ahead of time. That's why we have to do
             // it here.
-            let def_id = self.create_def(
-                node_id,
-                None,
-                DefKind::AnonConst,
-                DefPathData::LateAnonConst,
-                span,
-            );
+            let def_id = self.create_def(node_id, None, DefKind::AnonConst, span);
             let hir_id = self.lower_node_id(node_id);
 
             let path_expr = Expr {

--- a/compiler/rustc_ast_lowering/src/pat.rs
+++ b/compiler/rustc_ast_lowering/src/pat.rs
@@ -3,7 +3,6 @@ use std::sync::Arc;
 use rustc_ast::*;
 use rustc_data_structures::stack::ensure_sufficient_stack;
 use rustc_hir::def::{DefKind, Res};
-use rustc_hir::definitions::DefPathData;
 use rustc_hir::{self as hir, LangItem, Target};
 use rustc_middle::span_bug;
 use rustc_span::{DesugaringKind, Ident, Span, Spanned, respan};
@@ -538,8 +537,7 @@ impl<'hir, R: ResolverAstLoweringExt<'hir>> LoweringContext<'_, 'hir, R> {
         // We're generating a range end that didn't exist in the AST,
         // so the def collector didn't create the def ahead of time. That's why we have to do
         // it here.
-        let def_id =
-            self.create_def(node_id, None, DefKind::AnonConst, DefPathData::LateAnonConst, span);
+        let def_id = self.create_def(node_id, None, DefKind::AnonConst, span);
         let hir_id = self.lower_node_id(node_id);
 
         let unstable_span = self.mark_span_with_reason(

--- a/compiler/rustc_hir/src/definitions.rs
+++ b/compiler/rustc_hir/src/definitions.rs
@@ -97,9 +97,32 @@ impl DefPathTable {
     }
 }
 
-#[derive(Debug)]
+pub trait Disambiguator {
+    fn entry(&mut self, parent: LocalDefId, data: DefPathData) -> &mut u32;
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PerParentDisambiguatorState {
+    next: UnordMap<DefPathData, u32>,
+}
+
+impl Disambiguator for PerParentDisambiguatorState {
+    #[inline]
+    fn entry(&mut self, _: LocalDefId, data: DefPathData) -> &mut u32 {
+        self.next.entry(data).or_insert(0)
+    }
+}
+
+#[derive(Debug, Default, Clone)]
 pub struct DisambiguatorState {
     next: UnordMap<(LocalDefId, DefPathData), u32>,
+}
+
+impl Disambiguator for DisambiguatorState {
+    #[inline]
+    fn entry(&mut self, parent: LocalDefId, data: DefPathData) -> &mut u32 {
+        self.next.entry((parent, data)).or_insert(0)
+    }
 }
 
 impl DisambiguatorState {
@@ -302,10 +325,6 @@ pub enum DefPathData {
     Ctor,
     /// A constant expression (see `{ast,hir}::AnonConst`).
     AnonConst,
-    /// A constant expression created during AST->HIR lowering..
-    LateAnonConst,
-    /// A fresh anonymous lifetime created by desugaring elided lifetimes.
-    DesugaredAnonymousLifetime,
     /// An existential `impl Trait` type node.
     /// Argument position `impl Trait` have a `TypeNs` with their pretty-printed name.
     OpaqueTy,
@@ -389,7 +408,7 @@ impl Definitions {
         &mut self,
         parent: LocalDefId,
         data: DefPathData,
-        disambiguator: &mut DisambiguatorState,
+        disambiguator: &mut impl Disambiguator,
     ) -> LocalDefId {
         // We can't use `Debug` implementation for `LocalDefId` here, since it tries to acquire a
         // reference to `Definitions` and we're already holding a mutable reference.
@@ -403,7 +422,7 @@ impl Definitions {
 
         // Find the next free disambiguator for this key.
         let disambiguator = {
-            let next_disamb = disambiguator.next.entry((parent, data)).or_insert(0);
+            let next_disamb = disambiguator.entry(parent, data);
             let disambiguator = *next_disamb;
             *next_disamb = next_disamb.checked_add(1).expect("disambiguator overflow");
             disambiguator
@@ -458,8 +477,6 @@ impl DefPathData {
             TypeNs(name) | ValueNs(name) | MacroNs(name) | LifetimeNs(name)
             | OpaqueLifetime(name) => Some(name),
 
-            DesugaredAnonymousLifetime => Some(kw::UnderscoreLifetime),
-
             Impl
             | ForeignMod
             | CrateRoot
@@ -468,7 +485,6 @@ impl DefPathData {
             | Closure
             | Ctor
             | AnonConst
-            | LateAnonConst
             | OpaqueTy
             | AnonAssocTy(..)
             | SyntheticCoroutineBody
@@ -482,8 +498,6 @@ impl DefPathData {
             TypeNs(name) | ValueNs(name) | MacroNs(name) | LifetimeNs(name) | AnonAssocTy(name)
             | OpaqueLifetime(name) => Some(name),
 
-            DesugaredAnonymousLifetime => Some(kw::UnderscoreLifetime),
-
             Impl
             | ForeignMod
             | CrateRoot
@@ -492,7 +506,6 @@ impl DefPathData {
             | Closure
             | Ctor
             | AnonConst
-            | LateAnonConst
             | OpaqueTy
             | SyntheticCoroutineBody
             | NestedStatic => None,
@@ -512,8 +525,7 @@ impl DefPathData {
             GlobalAsm => DefPathDataName::Anon { namespace: sym::global_asm },
             Closure => DefPathDataName::Anon { namespace: sym::closure },
             Ctor => DefPathDataName::Anon { namespace: sym::constructor },
-            AnonConst | LateAnonConst => DefPathDataName::Anon { namespace: sym::constant },
-            DesugaredAnonymousLifetime => DefPathDataName::Named(kw::UnderscoreLifetime),
+            AnonConst => DefPathDataName::Anon { namespace: sym::constant },
             OpaqueTy => DefPathDataName::Anon { namespace: sym::opaque },
             AnonAssocTy(..) => DefPathDataName::Anon { namespace: sym::anon_assoc },
             SyntheticCoroutineBody => DefPathDataName::Anon { namespace: sym::synthetic },

--- a/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
+++ b/compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs
@@ -1523,7 +1523,7 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
                     None,
                     DefKind::LifetimeParam,
                     Some(DefPathData::OpaqueLifetime(ident.name)),
-                    &mut self.disambiguator,
+                    self.disambiguator,
                 );
                 feed.def_span(ident.span);
                 feed.def_ident_span(Some(ident.span));

--- a/compiler/rustc_middle/src/ty/context.rs
+++ b/compiler/rustc_middle/src/ty/context.rs
@@ -32,7 +32,7 @@ use rustc_data_structures::sync::{
 use rustc_errors::{Applicability, Diag, DiagCtxtHandle, Diagnostic, MultiSpan};
 use rustc_hir::def::DefKind;
 use rustc_hir::def_id::{CrateNum, DefId, LOCAL_CRATE, LocalDefId};
-use rustc_hir::definitions::{DefPathData, Definitions, DisambiguatorState};
+use rustc_hir::definitions::{DefPathData, Definitions, Disambiguator};
 use rustc_hir::intravisit::VisitorExt;
 use rustc_hir::lang_items::LangItem;
 use rustc_hir::limit::Limit;
@@ -1355,7 +1355,7 @@ impl<'tcx> TyCtxtAt<'tcx> {
         name: Option<Symbol>,
         def_kind: DefKind,
         override_def_path_data: Option<DefPathData>,
-        disambiguator: &mut DisambiguatorState,
+        disambiguator: &mut impl Disambiguator,
     ) -> TyCtxtFeed<'tcx, LocalDefId> {
         let feed =
             self.tcx.create_def(parent, name, def_kind, override_def_path_data, disambiguator);
@@ -1373,7 +1373,7 @@ impl<'tcx> TyCtxt<'tcx> {
         name: Option<Symbol>,
         def_kind: DefKind,
         override_def_path_data: Option<DefPathData>,
-        disambiguator: &mut DisambiguatorState,
+        disambiguator: &mut impl Disambiguator,
     ) -> TyCtxtFeed<'tcx, LocalDefId> {
         let data = override_def_path_data.unwrap_or_else(|| def_kind.def_path_data(name));
         // The following call has the side effect of modifying the tables inside `definitions`.

--- a/compiler/rustc_middle/src/ty/mod.rs
+++ b/compiler/rustc_middle/src/ty/mod.rs
@@ -40,6 +40,7 @@ use rustc_hir as hir;
 use rustc_hir::attrs::StrippedCfgItem;
 use rustc_hir::def::{CtorKind, CtorOf, DefKind, DocLinkResMap, LifetimeRes, Res};
 use rustc_hir::def_id::{CrateNum, DefId, DefIdMap, LocalDefId, LocalDefIdMap};
+use rustc_hir::definitions::PerParentDisambiguatorState;
 use rustc_hir::{LangItem, attrs as attr, find_attr};
 use rustc_index::IndexVec;
 use rustc_index::bit_set::BitMatrix;
@@ -224,6 +225,8 @@ pub struct ResolverAstLowering<'tcx> {
 
     // Information about delegations which is used when handling recursive delegations
     pub delegation_infos: LocalDefIdMap<DelegationInfo>,
+
+    pub per_parent_disambiguators: LocalDefIdMap<Steal<PerParentDisambiguatorState>>,
 }
 
 #[derive(Debug)]

--- a/compiler/rustc_middle/src/ty/print/pretty.rs
+++ b/compiler/rustc_middle/src/ty/print/pretty.rs
@@ -2172,7 +2172,6 @@ fn guess_def_namespace(tcx: TyCtxt<'_>, def_id: DefId) -> Namespace {
 
         DefPathData::ValueNs(..)
         | DefPathData::AnonConst
-        | DefPathData::LateAnonConst
         | DefPathData::Closure
         | DefPathData::Ctor => Namespace::ValueNS,
 

--- a/compiler/rustc_resolve/src/lib.rs
+++ b/compiler/rustc_resolve/src/lib.rs
@@ -57,7 +57,7 @@ use rustc_hir::def::{
     PerNS,
 };
 use rustc_hir::def_id::{CRATE_DEF_ID, CrateNum, DefId, LOCAL_CRATE, LocalDefId, LocalDefIdMap};
-use rustc_hir::definitions::DisambiguatorState;
+use rustc_hir::definitions::PerParentDisambiguatorState;
 use rustc_hir::{PrimTy, TraitCandidate, find_attr};
 use rustc_index::bit_set::DenseBitSet;
 use rustc_metadata::creader::CStore;
@@ -1353,7 +1353,7 @@ pub struct Resolver<'ra, 'tcx> {
 
     node_id_to_def_id: NodeMap<Feed<'tcx, LocalDefId>>,
 
-    disambiguator: DisambiguatorState = DisambiguatorState::new(),
+    per_parent_disambiguators: LocalDefIdMap<PerParentDisambiguatorState>,
 
     /// Indices of unnamed struct or variant fields with unresolved attributes.
     placeholder_field_indices: FxHashMap<NodeId, usize> = default::fx_hash_map(),
@@ -1557,7 +1557,13 @@ impl<'tcx> Resolver<'_, 'tcx> {
         );
 
         // FIXME: remove `def_span` body, pass in the right spans here and call `tcx.at().create_def()`
-        let feed = self.tcx.create_def(parent, name, def_kind, None, &mut self.disambiguator);
+        let feed = self.tcx.create_def(
+            parent,
+            name,
+            def_kind,
+            None,
+            self.per_parent_disambiguators.entry(parent).or_default(),
+        );
         let def_id = feed.def_id();
 
         // Create the definition.
@@ -1739,6 +1745,7 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             doc_link_resolutions: Default::default(),
             doc_link_traits_in_scope: Default::default(),
             current_crate_outer_attr_insert_span,
+            per_parent_disambiguators: Default::default(),
             ..
         };
 
@@ -1870,6 +1877,11 @@ impl<'ra, 'tcx> Resolver<'ra, 'tcx> {
             lifetime_elision_allowed: self.lifetime_elision_allowed,
             lint_buffer: Steal::new(self.lint_buffer),
             delegation_infos: self.delegation_infos,
+            per_parent_disambiguators: self
+                .per_parent_disambiguators
+                .into_items()
+                .map(|(k, d)| (k, Steal::new(d)))
+                .collect(),
         };
         ResolverOutputs { global_ctxt, ast_lowering }
     }

--- a/compiler/rustc_sanitizers/src/cfi/typeid/itanium_cxx_abi/encode.rs
+++ b/compiler/rustc_sanitizers/src/cfi/typeid/itanium_cxx_abi/encode.rs
@@ -680,7 +680,6 @@ fn encode_ty_name(tcx: TyCtxt<'_>, def_id: DefId) -> String {
             hir::definitions::DefPathData::Closure => "C",
             hir::definitions::DefPathData::Ctor => "c",
             hir::definitions::DefPathData::AnonConst => "K",
-            hir::definitions::DefPathData::LateAnonConst => "k",
             hir::definitions::DefPathData::OpaqueTy => "i",
             hir::definitions::DefPathData::SyntheticCoroutineBody => "s",
             hir::definitions::DefPathData::NestedStatic => "n",
@@ -690,7 +689,6 @@ fn encode_ty_name(tcx: TyCtxt<'_>, def_id: DefId) -> String {
             | hir::definitions::DefPathData::MacroNs(..)
             | hir::definitions::DefPathData::OpaqueLifetime(..)
             | hir::definitions::DefPathData::LifetimeNs(..)
-            | hir::definitions::DefPathData::DesugaredAnonymousLifetime
             | hir::definitions::DefPathData::AnonAssocTy(..) => {
                 bug!("encode_ty_name: unexpected `{:?}`", disambiguated_data.data);
             }

--- a/compiler/rustc_symbol_mangling/src/v0.rs
+++ b/compiler/rustc_symbol_mangling/src/v0.rs
@@ -888,7 +888,6 @@ impl<'tcx> Printer<'tcx> for V0SymbolMangler<'tcx> {
             DefPathData::Closure => 'C',
             DefPathData::Ctor => 'c',
             DefPathData::AnonConst => 'K',
-            DefPathData::LateAnonConst => 'k',
             DefPathData::OpaqueTy => 'i',
             DefPathData::SyntheticCoroutineBody => 's',
             DefPathData::NestedStatic => 'n',
@@ -900,7 +899,6 @@ impl<'tcx> Printer<'tcx> for V0SymbolMangler<'tcx> {
             | DefPathData::Impl
             | DefPathData::MacroNs(_)
             | DefPathData::LifetimeNs(_)
-            | DefPathData::DesugaredAnonymousLifetime
             | DefPathData::OpaqueLifetime(_)
             | DefPathData::AnonAssocTy(..) => {
                 bug!("symbol_names: unexpected DefPathData: {:?}", disambiguated_data.data)

--- a/tests/ui/delegation/generics/def-path-hash-collision-ice-153410.rs
+++ b/tests/ui/delegation/generics/def-path-hash-collision-ice-153410.rs
@@ -1,0 +1,27 @@
+//@ compile-flags: -Z deduplicate-diagnostics=yes
+//@ edition:2024
+
+#![feature(fn_delegation)]
+#![feature(iter_advance_by)]
+#![feature(iter_array_chunks)]
+#![feature(iterator_try_collect)]
+#![feature(iterator_try_reduce)]
+#![feature(iter_collect_into)]
+#![feature(iter_intersperse)]
+#![feature(iter_is_partitioned)]
+#![feature(iter_map_windows)]
+#![feature(iter_next_chunk)]
+#![feature(iter_order_by)]
+#![feature(iter_partition_in_place)]
+#![feature(trusted_random_access)]
+#![feature(try_find)]
+#![allow(incomplete_features)]
+
+impl Iterator {
+//~^ ERROR: expected a type, found a trait [E0782]
+    reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
+    //~^ ERROR: expected method or associated constant, found associated type `Iterator::Item`
+    //~| ERROR: ambiguous associated type
+}
+
+fn main() {}

--- a/tests/ui/delegation/generics/def-path-hash-collision-ice-153410.stderr
+++ b/tests/ui/delegation/generics/def-path-hash-collision-ice-153410.stderr
@@ -1,0 +1,40 @@
+error[E0575]: expected method or associated constant, found associated type `Iterator::Item`
+  --> $DIR/def-path-hash-collision-ice-153410.rs:22:10
+   |
+LL |     reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
+   |          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ not a method or associated constant
+
+error[E0782]: expected a type, found a trait
+  --> $DIR/def-path-hash-collision-ice-153410.rs:20:6
+   |
+LL | impl Iterator {
+   |      ^^^^^^^^
+   |
+help: you can add the `dyn` keyword if you want a trait object
+   |
+LL | impl dyn Iterator {
+   |      +++
+help: you might have intended to implement this trait for a given type
+   |
+LL | impl Iterator for /* Type */ {
+   |               ++++++++++++++
+
+error[E0223]: ambiguous associated type
+  --> $DIR/def-path-hash-collision-ice-153410.rs:22:14
+   |
+LL |     reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
+   |              ^^^^^^^^^^^^^^^^^^^^^^
+   |
+help: use fully-qualified syntax
+   |
+LL -     reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
+LL +     reuse< < <fn() as BitOr>::Output>::Item as Iterator>::*;
+   |
+LL -     reuse< < <for<'a> fn()>::Output>::Item as Iterator>::*;
+LL +     reuse< < <fn() as IntoFuture>::Output>::Item as Iterator>::*;
+   |
+
+error: aborting due to 3 previous errors
+
+Some errors have detailed explanations: E0223, E0575, E0782.
+For more information about an error, try `rustc --explain E0223`.

